### DISCUSSION TITLE
Best practices of define the boundary.

### DISCUSSION THREAD

**Discussion by xfdywy** [firedrakeproject/firedrake#2860] [Firedrake support] [answered]
Hello, everyone. If the boundary is complex,  could anyone tell me what is the best practice of defining the boundary?  For example, considering the case I need to distinguish the left, right, top, and bottom boundary. Can I use a python function to define the different boundary just like in fenics:
```
# Boundaries
def right(x, on_boundary): return x[0] > (1.0 - DOLFIN_EPS)
def left(x, on_boundary): return x[0] < DOLFIN_EPS
def top_bottom(x, on_boundary):
    return x[1] > 1.0 - DOLFIN_EPS or x[1] < DOLFIN_EPS

# No-slip boundary condition for velocity
noslip = Constant((0.0, 0.0, 0.0))
bc0 = DirichletBC(W.sub(0), noslip, top_bottom)

# Inflow boundary condition for velocity
inflow = Expression(("-sin(x[1]*pi)", "0.0", "0.0"), degree=2)
bc1 = DirichletBC(W.sub(0), inflow, right)
```
Or in this way to modified the facet markers directly like in fenics:
 ```
   bndry = MeshFunction("size_t", mesh, mesh.topology().dim()-1)
    for f in facets(mesh):
        mp = f.midpoint()
        if near(mp[0], 0.0):  # inflow
            bndry[f] = 1
        elif near(mp[0], L):  # outflow
            bndry[f] = 2
        elif near(mp[1], 0.0) or near(mp[1], W):  # walls
            bndry[f] = 3
        elif mp.distance(center) <= radius:  # cylinder
            bndry[f] = 5
```
Or in this way like in fenicsx:
```
def walls(x):
    return np.logical_or(np.isclose(x[1],0), np.isclose(x[1],1))
wall_dofs = locate_dofs_geometrical(V, walls)
u_noslip = np.array((0,) * mesh.geometry.dim, dtype=PETSc.ScalarType)
bc_noslip = dirichletbc(u_noslip, wall_dofs, V)
```
Thanks~

**Comment by ksagiyam** [comment, marked as answer]
`RelabeledMesh` (https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.RelabeledMesh) lets you make a new mesh with new labels; see https://github.com/firedrakeproject/firedrake/blob/master/tests/regression/test_mark_entities.py for examples.

**Reply by enricofacca** [reply]
Hello, 
I am reading this discussion because I am trying to label some faces using an indicator function **but only on the boundary** in order to define some Dirichlet boundary condition. 
It is there way to do it using RelabeledMesh or other existing Firedrake tools without manipulating directly the dm_topolgy object?
Thank you.
Enrico

**Reply by connorjward** [reply]
This discussion is very old. Can you please open a new one to ask this?

**Reply by enricofacca** [reply]
yes, sure.
